### ISSUE TITLE
DAPO loss is not normalized by gradient_accumulation_steps

### ISSUE THREAD

**Issue by haotang1995** [huggingface/trl#5619]
Hi, I noticed that the DAPO loss is not normalized by current_gradient_accumulation_steps and am not sure why.  

```

        if self.loss_type in ["grpo", "sapo"]:
            loss = ((per_token_loss * mask).sum(-1) / mask.sum(-1).clamp(min=1.0)).mean()
            normalizer = self.current_gradient_accumulation_steps if mode == "train" else 1.0  # no accum in eval
            loss = loss / normalizer
        elif self.loss_type in ["cispo", "dapo", "vespo"]:
            normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
            loss = (per_token_loss * mask).sum() / normalizer

```
in https://github.com/huggingface/trl/blob/a08e7139f933b770177fc2abc0b43118e26b260b/trl/trainer/grpo_trainer.py#L2547-L2562

**Comment by qgallouedec** [comment]
`inputs["num_items_in_batch"]` corresponds to the number of items in the accumulated batch, so technically yes, gradient accumulation is taken into account

**Comment by haotang1995** [comment]
Thanks for the reply. May I ask if num_items_in_batch  is correlated with steps_per_generation or graidient_accumulation_steps? I feel like they are different.

**Comment by Adya-Prasad** [comment]
Hi @qgallouedec  I'm Adya. I have been contributing to HF Peft, and now I want to contribute to TRL. Can I pick up this issue and raise PR ? OR you can suggest other acute issue to pickup (easy to medium level)

**Comment by 0xadvait** [comment]
The two are only equivalent when they coincide. `num_items_in_batch` is computed over the generation batch, which spans `steps_per_generation` micro-steps, while one optimizer update only accumulates `gradient_accumulation_steps` micro-steps. So the accumulated DAPO/CISPO/VESPO loss is scaled by `gradient_accumulation_steps / steps_per_generation` whenever the two differ.

Measured with a spy on `_compute_loss` (tiny model, `loss_type="dapo"`, `num_iterations=1`, `beta=0`, so the expected loss is exactly reconstructable from the inputs):

- `steps_per_generation=4, gradient_accumulation_steps=2`: accumulated window loss is 0.500x the token-mean over the window
- `steps_per_generation=2, gradient_accumulation_steps=2`: 1.000x (default-equivalent, correct)
- `steps_per_generation=2, gradient_accumulation_steps=4`: 2.0x

The Lite PPO recipe in the paper index (`loss_type="dapo"`, `gradient_accumulation_steps=32`, `steps_per_generation=8`) is scaled 4x by this.

Opened #6024 which rescales the normalizer to the accumulation window token count (no-op when the two values coincide).
